crates/bevy_ecs/src/component.rs

```diff
@@ -279,42 +269,58 @@ impl Components {
 
     #[inline]
     pub fn get_or_insert_resource_id<T: Component>(&mut self) -> ComponentId {
-        self.get_or_insert_resource_with(TypeId::of::<T>(), TypeInfo::of::<T>)
+        // SAFE: The [`ComponentDescriptor`] matches the [`TypeId`]
+        unsafe {
+            self.get_or_insert_resource_with(TypeId::of::<T>(), || {
+                ComponentDescriptor::new::<T>(StorageType::default())
+            })
+        }
     }
 
     #[inline]
     pub fn get_or_insert_non_send_resource_id<T: Any>(&mut self) -> ComponentId {
-        self.get_or_insert_resource_with(TypeId::of::<T>(), TypeInfo::of_non_send_and_sync::<T>)
+        // SAFE: The [`ComponentDescriptor`] matches the [`TypeId`]
+        unsafe {
+            self.get_or_insert_resource_with(TypeId::of::<T>(), || {
+                ComponentDescriptor::new_non_send::<T>(StorageType::default())
+            })
+        }
     }
 
+    /// # Safety
+    ///
+    /// The [`ComponentDescriptor`] must match the [`TypeId`]
     #[inline]
-    fn get_or_insert_resource_with(
+    unsafe fn get_or_insert_resource_with(
         &mut self,
         type_id: TypeId,
-        func: impl FnOnce() -> TypeInfo,
+        func: impl FnOnce() -> ComponentDescriptor,
```

Comment:
do we know if perf is actually worse without this `func` param? can we just take a `ComponentDescriptor` always and drop the `type_id` arg?

Reply:
`ComponentDescriptor` allocates a `String` for the type name. That is not something you want in the hot path.

Reply:
ah right, the string... rip

Reply:
type_name returns a `&'static str` though, seems like we could atleast use `Cow<&'static str>` here. Wouldn't be the best for custom names though :frown:

Reply:
`Cow<'static, str>` would work. The name is not frequently used, so the extra branch shouldn't hurt. I will leave that for another PR though I think.

Reply:
I think personally I'd rather drop the unsafe, use a Cow and then always take a `ComponentDescriptor`

Reply:
`ComponentDescriptor` is 72 bytes with `String` and 80 bytes with Cow, which is relatively large. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=93f494d4a982196b786f0f7d7f69c895

Reply:
I dont know if that actually tells us anything about the performance here though?